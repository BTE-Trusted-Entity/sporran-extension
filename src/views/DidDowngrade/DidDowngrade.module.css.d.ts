export const __esModule: true;
export const container: string;
export const heading: string;
export const subline: string;
export const costs: string;
export const details: string;
export const info: string;
export const buttonsLine: string;
export const cancel: string;
export const submit: string;
export const errorTooltip: string;
export const explanation: string;
