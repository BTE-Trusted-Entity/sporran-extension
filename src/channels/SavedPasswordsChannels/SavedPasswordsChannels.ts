import { BrowserChannel } from '../base/BrowserChannel/BrowserChannel';
import { channelsEnum } from '../base/channelsEnum';

interface SavedPassword {
  password: string;
  timestamp: number;
}

const savedPasswords: Record<string, SavedPassword> = {};

interface SavePasswordInput {
  password: string;
  address: string;
}

export const savePasswordChannel = new BrowserChannel<SavePasswordInput>(
  channelsEnum.savePassword,
);

export async function savePassword({
  address,
  password,
}: SavePasswordInput): Promise<void> {
  savedPasswords[address] = { password, timestamp: Date.now() };
}

export const getPasswordChannel = new BrowserChannel<
  string,
  string | undefined
>(channelsEnum.getPassword);

export async function getPassword(
  address: string,
): Promise<string | undefined> {
  return savedPasswords[address]?.password;
}

export const forgetPasswordChannel = new BrowserChannel<string>(
  channelsEnum.forgetPassword,
);

export async function forgetPassword(address: string): Promise<void> {
  delete savedPasswords[address];
}

export const hasSavedPasswordsChannel = new BrowserChannel<void, boolean>(
  channelsEnum.hasSavedPasswords,
);

export async function hasSavedPasswords(): Promise<boolean> {
  return Object.values(savedPasswords).length > 0;
}

export const forgetAllPasswordsChannel = new BrowserChannel(
  channelsEnum.forgetAllPasswords,
);

export async function forgetAllPasswords(): Promise<void> {
  for (const password in savedPasswords) {
    delete savedPasswords[password];
  }
}

const saveDuration = 15 * 60 * 1000;
const intervalDuration = 1 * 60 * 1000;

function checkExpiredPasswords(): void {
  const oldestPossible = Date.now() - saveDuration;
  for (const password in savedPasswords) {
    const { timestamp } = savedPasswords[password];
    if (timestamp < oldestPossible) {
      delete savedPasswords[password];
    }
  }
}

export function schedulePasswordsCheck(): void {
  setInterval(checkExpiredPasswords, intervalDuration);
}
