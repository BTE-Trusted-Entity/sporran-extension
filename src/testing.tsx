export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
