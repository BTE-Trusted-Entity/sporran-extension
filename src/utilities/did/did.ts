import {
  ConfigService,
  Did,
  DidDocument,
  DidEncryptionKey,
  DidUri,
  Utils,
} from '@kiltprotocol/sdk-js';

import { useAsyncValue } from '../useAsyncValue/useAsyncValue';

export function isFullDid(did: DidUri | undefined): boolean {
  if (!did) {
    // could be a legacy identity without DID
    return false;
  }
  return Did.parse(did).type === 'full';
}

export async function getFullDidDocument(did: DidUri): Promise<DidDocument> {
  const api = ConfigService.get('api');
  return Did.linkedInfoFromChain(await api.call.did.query(Did.toChain(did)))
    .document;
}

export function useFullDidDocument(did: DidUri): DidDocument | undefined {
  return useAsyncValue(getFullDidDocument, [did]);
}

export async function getDidDocument(did: DidUri): Promise<DidDocument> {
  return isFullDid(did)
    ? await getFullDidDocument(did)
    : Did.parseDocumentFromLightDid(did);
}

export function getDidEncryptionKey({
  keyAgreement,
}: DidDocument): DidEncryptionKey {
  if (!keyAgreement || !keyAgreement[0]) {
    throw new Error('encryptionKey is not defined somehow');
  }
  return keyAgreement[0];
}

export function parseDidUri(did: DidUri): ReturnType<typeof Did.parse> & {
  fullDid: DidUri;
} {
  const parsed = Did.parse(did);
  const fullDid = parsed.type === 'full' ? did : Did.getFullDidUri(did);

  return {
    ...parsed,
    fullDid,
  };
}

export async function needLegacyDidCrypto(
  did: DidUri | undefined,
): Promise<boolean> {
  if (!did) {
    // DID was deactivated, no action needed.
    return false;
  }
  if (!isFullDid(did)) {
    return false;
  }

  try {
    const encryptionKey = getDidEncryptionKey(await getDidDocument(did));
    return (
      Utils.Crypto.u8aToHex(encryptionKey.publicKey) ===
      '0xf2c90875e0630bd1700412341e5e9339a57d2fefdbba08de1cac8db5b4145f6e'
    );
  } catch {
    // getDidDocument might throw if the DID is not on-chain anymore (removed, another endpoint),
    // no legacy crypto needed in that case
    return false;
  }
}
