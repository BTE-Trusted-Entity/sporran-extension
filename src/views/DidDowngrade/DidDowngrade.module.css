.container {
  composes: container from '../../components/View/View.module.css';
}

.heading {
  composes: screenHeadline from '../../components/Typography/Typography.module.css';
}

.subline {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin: 0 0 0.8125rem 0;
}

.costs {
  composes: smallText from '../../components/Typography/Typography.module.css';
  font-size: 1.25rem;
  margin: 0.25rem 0 0 0;
  display: flex;
  align-items: center;
  justify-content: center;
}

.details {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin: 0.5rem 0 1.5rem 0;
}

.info {
  background: url('../../components/images/info-solid.svg') no-repeat center/100%;
  display: inline-block;
  width: 1em;
  height: 1em;
  vertical-align: text-top;
  margin-top: 0.125em;
}

.buttonsLine {
  margin: 1.5rem 0 0;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 1rem;
  position: relative;
}

.cancel {
  composes: buttonSecondary from '../../components/Button/Button.module.css';
}

.submit {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
}

.errorTooltip {
  composes: tooltip from '../../components/ErrorTooltip/ErrorTooltip.module.css';
  top: 2.5rem;
  left: 75%;
}
