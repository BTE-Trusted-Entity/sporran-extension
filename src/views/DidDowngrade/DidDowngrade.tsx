import { FormEvent, Fragment, useCallback, useMemo, useState } from 'react';
import { Link } from 'react-router-dom';
import BN from 'bn.js';
import { browser } from 'webextension-polyfill-ts';
import { DidUri } from '@kiltprotocol/sdk-js';

import * as styles from './DidDowngrade.module.css';

import { Identity } from '../../utilities/identities/types';
import {
  getFee,
  sign,
  submit,
} from '../../utilities/didDowngrade/didDowngrade';
import { saveIdentity } from '../../utilities/identities/identities';
import { paths } from '../paths';

import { IdentitySlide } from '../../components/IdentitySlide/IdentitySlide';
import {
  PasswordError,
  PasswordField,
  usePasswordField,
} from '../../components/PasswordField/PasswordField';
import { useAddressBalance } from '../../components/Balance/Balance';
import { KiltCurrency } from '../../components/KiltCurrency/KiltCurrency';
import {
  asKiltCoins,
  KiltAmount,
} from '../../components/KiltAmount/KiltAmount';
import { TxStatusModal } from '../../components/TxStatusModal/TxStatusModal';
import { LinkBack } from '../../components/LinkBack/LinkBack';
import { Stats } from '../../components/Stats/Stats';
import {
  invalidateCredentials,
  useIdentityCredentials,
} from '../../utilities/credentials/credentials';
import {
  useDepositDid,
  useDepositWeb3Name,
} from '../../utilities/getDeposit/getDeposit';
import { useAsyncValue } from '../../utilities/useAsyncValue/useAsyncValue';

function useCosts(
  address: string,
  did: DidUri | undefined, // after the downgrade itâ€™s set to undefined
): {
  fee?: BN;
  deposit?: BN;
  total?: BN;
  error: boolean;
} {
  const fee = useAsyncValue(getFee, [did]);
  const depositDid = useDepositDid(did);
  const depositWeb3Name = useDepositWeb3Name(did);

  const deposit = useMemo(() => {
    const didAmount =
      depositDid?.owner === address ? depositDid.amount : new BN(0);
    const w3nAmount =
      depositWeb3Name?.owner === address ? depositWeb3Name.amount : new BN(0);
    return didAmount.add(w3nAmount);
  }, [address, depositDid, depositWeb3Name]);

  const total = useMemo(
    () => (fee && !deposit.isZero() ? deposit.sub(fee) : undefined),
    [fee, deposit],
  );

  const balance = useAddressBalance(address);
  const error = Boolean(fee && balance && balance.transferable.lt(fee));

  return { fee, deposit, total, error };
}

interface Props {
  identity: Identity;
}

export function DidDowngrade({ identity }: Props): JSX.Element | null {
  const t = browser.i18n.getMessage;

  const { address, did } = identity;
  const { fee, deposit, total, error } = useCosts(address, did);
  const [txHash, setTxHash] = useState<string>();

  const [submitting, setSubmitting] = useState(false);
  const [status, setStatus] = useState<'pending' | 'success' | 'error' | null>(
    null,
  );

  const sporranCredentials = useIdentityCredentials(did);

  const passwordField = usePasswordField();
  const handleSubmit = useCallback(
    async (event: FormEvent) => {
      event.preventDefault();

      try {
        const { seed } = await passwordField.get(event);

        setSubmitting(true);
        setStatus('pending');

        const hash = await sign(identity, seed);
        setTxHash(hash);

        await submit(hash);
        await saveIdentity({
          ...identity,
          did: undefined,
          deletedDid: identity.did,
        });

        if (sporranCredentials) {
          await invalidateCredentials(sporranCredentials);
        }

        setStatus('success');
      } catch (error) {
        if (error instanceof PasswordError) {
          return;
        }
        console.error(error);
        setSubmitting(false);
        setStatus('error');
      }
    },
    [identity, passwordField, sporranCredentials],
  );

  const closeModal = useCallback(() => {
    setStatus(null);
  }, []);

  if (!fee) {
    return null; // blockchain data pending
  }

  if (!sporranCredentials) {
    return null; // storage data pending
  }

  return (
    <form
      onSubmit={handleSubmit}
      className={styles.container}
      autoComplete="off"
    >
      <h1 className={styles.heading}>{t('view_DidDowngrade_heading')}</h1>
      <p className={styles.subline}>{t('view_DidDowngrade_subline')}</p>

      <IdentitySlide identity={identity} options={false} />

      {total && deposit && (
        <Fragment>
          <p className={styles.costs}>
            {t('view_DidDowngrade_total')}
            <KiltAmount amount={total} type="costs" smallDecimals />
          </p>
          <p className={styles.details}>
            {t('view_DidDowngrade_deposit')}
            {asKiltCoins(deposit, 'funds')} <KiltCurrency />
            {t('view_DidDowngrade_fee')}
            {asKiltCoins(fee, 'costs')} <KiltCurrency />
          </p>
        </Fragment>
      )}

      {!total && (
        <p className={styles.costs}>
          {t('view_DidDowngrade_fee_as_total')}
          <KiltAmount amount={fee} type="costs" smallDecimals />
        </p>
      )}

      <p className={styles.explanation}>{t('view_DidDowngrade_explanation')}</p>

      <PasswordField identity={identity} autoFocus password={passwordField} />

      <p className={styles.buttonsLine}>
        <Link to={paths.home} className={styles.cancel}>
          {t('common_action_cancel')}
        </Link>
        <button
          type="submit"
          className={styles.submit}
          disabled={submitting || error}
        >
          {t('common_action_sign')}
        </button>
        <output
          className={styles.errorTooltip}
          hidden={!error || Boolean(status)}
        >
          {t('view_DidDowngrade_insufficientFunds', [
            asKiltCoins(fee, 'costs'),
          ])}
        </output>
      </p>

      {status && (
        <TxStatusModal
          identity={identity}
          status={status}
          txHash={txHash}
          onDismissError={closeModal}
          messages={{
            pending: t('view_DidDowngrade_pending'),
            success: t('view_DidDowngrade_success'),
            error: t('view_DidDowngrade_error'),
          }}
        />
      )}

      <LinkBack />
      <Stats />
    </form>
  );
}
