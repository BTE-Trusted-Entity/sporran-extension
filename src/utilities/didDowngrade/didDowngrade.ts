import BN from 'bn.js';
import {
  Blockchain,
  ConfigService,
  Did,
  DidUri,
  KiltKeyringPair,
  SignExtrinsicCallback,
  SubmittableExtrinsic,
} from '@kiltprotocol/sdk-js';

import { getIdentityCryptoFromSeed, Identity } from '../identities/identities';
import { isFullDid } from '../did/did';
import { makeFakeIdentityCrypto } from '../makeFakeIdentityCrypto/makeFakeIdentityCrypto';

interface DidTransaction {
  extrinsic: SubmittableExtrinsic;
}

async function getSignedTransaction(
  keypair: KiltKeyringPair,
  sign: SignExtrinsicCallback,
  did: DidUri,
): Promise<DidTransaction> {
  const api = ConfigService.get('api');
  const submitter = keypair.address;

  const chainDid = Did.toChain(did);
  const { web3Name, document } = Did.linkedInfoFromChain(
    await api.call.did.query(chainDid),
  );
  const servicesCount = document.service?.length ?? 0;

  let authorized: SubmittableExtrinsic;

  if (web3Name) {
    authorized = await Did.authorizeBatch({
      batchFunction: api.tx.utility.batchAll,
      did,
      extrinsics: [
        api.tx.web3Names.releaseByOwner(),
        api.tx.did.delete(servicesCount),
      ],
      sign,
      submitter,
    });
  } else {
    authorized = await Did.authorizeTx(
      document.uri,
      api.tx.did.delete(servicesCount),
      sign,
      submitter,
    );
  }

  const extrinsic = await authorized.signAsync(keypair);

  return { extrinsic };
}

export async function getFee(did: DidUri | undefined): Promise<BN> {
  if (!did || !isFullDid(did)) {
    return new BN(0);
  }
  const { keypair, sign } = makeFakeIdentityCrypto();
  const { extrinsic } = await getSignedTransaction(keypair, sign, did);

  return (await extrinsic.paymentInfo(keypair)).partialFee;
}

const currentTx: Record<string, DidTransaction> = {};

export async function sign(
  identity: Identity,
  seed: Uint8Array,
): Promise<string> {
  if (!identity.did) {
    throw new Error('DID is deleted and unusable');
  }
  const { keypair, sign } = await getIdentityCryptoFromSeed(seed);
  const { extrinsic } = await getSignedTransaction(keypair, sign, identity.did);

  const hash = extrinsic.hash.toHex();
  currentTx[hash] = { extrinsic };
  return hash;
}

export async function submit(hash: string): Promise<void> {
  const { extrinsic } = currentTx[hash];
  await Blockchain.submitSignedTx(extrinsic);
  delete currentTx[hash];
}
