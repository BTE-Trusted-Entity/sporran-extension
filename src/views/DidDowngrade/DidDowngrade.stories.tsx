import { Meta } from '@storybook/react';

import { moreIdentitiesMock as identities } from '../../utilities/identities/IdentitiesProvider.mock';

import { DidDowngrade } from './DidDowngrade';

export default {
  title: 'Views/DidDowngrade',
  component: DidDowngrade,
} as Meta;

export function DidRefund(): JSX.Element {
  return (
    <DidDowngrade
      identity={identities['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo']}
    />
  );
}

export function W3NRefund(): JSX.Element {
  return (
    <DidDowngrade
      identity={identities['4p273cfeZ2JRz46AcJoQvTRHCH8Vaj92jts2VxepZtQwbTBB']}
    />
  );
}

export function DidAndW3NRefund(): JSX.Element {
  return (
    <DidDowngrade
      identity={identities['4q11Jce9wqM4A9GPB2z8n4K8LF9w2sQgZKFddhuKXwQ2Qo4q']}
    />
  );
}

export function NoRefund(): JSX.Element {
  return (
    <DidDowngrade
      identity={identities['4su6rRjEVPfNYCuaXw7iF3os1REHE6Gan23mYo2vc6fT7jZq']}
    />
  );
}
