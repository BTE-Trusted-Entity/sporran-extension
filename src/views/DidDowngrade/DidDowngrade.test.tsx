import { BalanceUtils } from '@kiltprotocol/sdk-js';

import { identitiesMock as identities, render } from '../../testing/testing';

import { waitForGetPassword } from '../../channels/SavedPasswordsChannels/SavedPasswordsChannels.mock';
import { getFee } from '../../utilities/didDowngrade/didDowngrade';
import { mockIsFullDid } from '../../utilities/did/did.mock';

import { parseDidUri } from '../../utilities/did/did';

import {
  useDepositDid,
  useDepositWeb3Name,
} from '../../utilities/getDeposit/getDeposit';

import { DidDowngrade } from './DidDowngrade';

jest.mock('../../utilities/didDowngrade/didDowngrade');
jest.mocked(getFee).mockResolvedValue(BalanceUtils.toFemtoKilt(0.01));

jest.mock('../../utilities/did/did');
jest.mocked(parseDidUri).mockReturnValue({
  fullDid: 'did:kilt:4rrkiRTZgsgxjJDFkLsivqqKTqdUTuxKk3FX3mKFAeMxsR51',
} as unknown as ReturnType<typeof parseDidUri>);

jest.mock('../../utilities/getDeposit/getDeposit');

const depositAmount = BalanceUtils.toFemtoKilt(1);

describe('DidDowngrade', () => {
  it('promo used for both web3name and DID', async () => {
    mockIsFullDid(true);
    jest.mocked(useDepositDid).mockReturnValue({
      owner: '4promo account',
      amount: depositAmount,
    });
    jest.mocked(useDepositWeb3Name).mockReturnValue({
      owner: '4promo account',
      amount: depositAmount,
    });

    const { container } = render(
      <DidDowngrade
        identity={
          identities['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo']
        }
      />,
    );
    await waitForGetPassword();
    expect(container).toMatchSnapshot();
  });

  it('promo used for web3name but not DID', async () => {
    mockIsFullDid(true);
    jest.mocked(useDepositDid).mockReturnValue({
      owner: '4promo account',
      amount: depositAmount,
    });
    jest.mocked(useDepositWeb3Name).mockReturnValue({
      owner: '4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo',
      amount: depositAmount,
    });

    const { container } = render(
      <DidDowngrade
        identity={
          identities['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo']
        }
      />,
    );
    await waitForGetPassword();
    expect(container).toMatchSnapshot();
  });
  it('promo used for DID but not web3name', async () => {
    mockIsFullDid(true);
    jest.mocked(useDepositDid).mockReturnValue({
      owner: '4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo',
      amount: depositAmount,
    });
    jest.mocked(useDepositWeb3Name).mockReturnValue({
      owner: '4promo account',
      amount: depositAmount,
    });

    const { container } = render(
      <DidDowngrade
        identity={
          identities['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo']
        }
      />,
    );
    await waitForGetPassword();
    expect(container).toMatchSnapshot();
  });
  it('promo not used at all', async () => {
    mockIsFullDid(true);
    jest.mocked(useDepositDid).mockReturnValue({
      owner: '4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo',
      amount: depositAmount,
    });
    jest.mocked(useDepositWeb3Name).mockReturnValue({
      owner: '4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo',
      amount: depositAmount,
    });

    const { container } = render(
      <DidDowngrade
        identity={
          identities['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo']
        }
      />,
    );
    await waitForGetPassword();
    expect(container).toMatchSnapshot();
  });
});
