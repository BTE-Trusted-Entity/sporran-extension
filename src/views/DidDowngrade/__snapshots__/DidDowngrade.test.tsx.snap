// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DidDowngrade promo not used at all 1`] = `
<div>
  <form
    autocomplete="off"
    class="container"
  >
    <h1
      class="heading"
    >
      Remove DID
    </h1>
    <p
      class="subline"
    >
      Remove DID from the KILT blockchain
    </p>
    <section
      class="container"
    >
      <div
        class="avatar fullDid"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Identity with on-chain DID and lots of coins
        </span>
      </div>
    </section>
    <p
      class="costs"
    >
      Refund Amount: 
      <span
        class="amount"
      >
        1
        <span
          class="fractional"
        >
          .9900
        </span>
         
        <span
          aria-label="KILT Coin"
          class="coin"
        />
      </span>
    </p>
    <p
      class="details"
    >
       Deposit 
      2.0000
       
      <span
        aria-label="KILT Coin"
        class="coin"
      />
       - fee 
      0.0100
       
      <span
        aria-label="KILT Coin"
        class="coin"
      />
    </p>
    <p
      class="explanation"
    >
      Take care: this identity won’t be able to get a DID again.
    </p>
    <p
      class="resetLine"
    >
      <label
        class="passwordLabel"
        for="password"
      >
        Sign with password:
      </label>
      <a
        class="reset"
        href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo/reset"
      >
        Reset password
      </a>
    </p>
    <p
      class="passwordLine"
    >
      <input
        class="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        aria-label="Show"
        class="show"
        title="Show"
        type="button"
      />
      <output
        class="errorTooltip"
        hidden=""
      />
    </p>
    <label
      class="rememberLabel"
    >
      <span>
        Remember password for the next 15 minutes
      </span>
      <input
        class="remember"
        name="remember"
        type="checkbox"
      />
      <span />
    </label>
    <p
      class="buttonsLine"
    >
      <a
        class="cancel"
        href="/"
      >
        Cancel
      </a>
      <button
        class="submit"
        type="submit"
      >
        Sign
      </button>
      <output
        class="errorTooltip"
        hidden=""
      >
        You need at least 0.0100 KILTs to pay the fee
      </output>
    </p>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </form>
</div>
`;

exports[`DidDowngrade promo used for DID but not web3name 1`] = `
<div>
  <form
    autocomplete="off"
    class="container"
  >
    <h1
      class="heading"
    >
      Remove DID
    </h1>
    <p
      class="subline"
    >
      Remove DID from the KILT blockchain
    </p>
    <section
      class="container"
    >
      <div
        class="avatar fullDid"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Identity with on-chain DID and lots of coins
        </span>
      </div>
    </section>
    <p
      class="costs"
    >
      Refund Amount: 
      <span
        class="amount"
      >
        0
        <span
          class="fractional"
        >
          .9900
        </span>
         
        <span
          aria-label="KILT Coin"
          class="coin"
        />
      </span>
    </p>
    <p
      class="details"
    >
       Deposit 
      1.0000
       
      <span
        aria-label="KILT Coin"
        class="coin"
      />
       - fee 
      0.0100
       
      <span
        aria-label="KILT Coin"
        class="coin"
      />
    </p>
    <p
      class="explanation"
    >
      Take care: this identity won’t be able to get a DID again.
    </p>
    <p
      class="resetLine"
    >
      <label
        class="passwordLabel"
        for="password"
      >
        Sign with password:
      </label>
      <a
        class="reset"
        href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo/reset"
      >
        Reset password
      </a>
    </p>
    <p
      class="passwordLine"
    >
      <input
        class="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        aria-label="Show"
        class="show"
        title="Show"
        type="button"
      />
      <output
        class="errorTooltip"
        hidden=""
      />
    </p>
    <label
      class="rememberLabel"
    >
      <span>
        Remember password for the next 15 minutes
      </span>
      <input
        class="remember"
        name="remember"
        type="checkbox"
      />
      <span />
    </label>
    <p
      class="buttonsLine"
    >
      <a
        class="cancel"
        href="/"
      >
        Cancel
      </a>
      <button
        class="submit"
        type="submit"
      >
        Sign
      </button>
      <output
        class="errorTooltip"
        hidden=""
      >
        You need at least 0.0100 KILTs to pay the fee
      </output>
    </p>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </form>
</div>
`;

exports[`DidDowngrade promo used for both web3name and DID 1`] = `
<div>
  <form
    autocomplete="off"
    class="container"
  >
    <h1
      class="heading"
    >
      Remove DID
    </h1>
    <p
      class="subline"
    >
      Remove DID from the KILT blockchain
    </p>
    <section
      class="container"
    >
      <div
        class="avatar fullDid"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Identity with on-chain DID and lots of coins
        </span>
      </div>
    </section>
    <p
      class="costs"
    >
      Fee: 
      <span
        class="amount"
      >
        0
        <span
          class="fractional"
        >
          .0100
        </span>
         
        <span
          aria-label="KILT Coin"
          class="coin"
        />
      </span>
    </p>
    <p
      class="explanation"
    >
      Take care: this identity won’t be able to get a DID again.
    </p>
    <p
      class="resetLine"
    >
      <label
        class="passwordLabel"
        for="password"
      >
        Sign with password:
      </label>
      <a
        class="reset"
        href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo/reset"
      >
        Reset password
      </a>
    </p>
    <p
      class="passwordLine"
    >
      <input
        class="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        aria-label="Show"
        class="show"
        title="Show"
        type="button"
      />
      <output
        class="errorTooltip"
        hidden=""
      />
    </p>
    <label
      class="rememberLabel"
    >
      <span>
        Remember password for the next 15 minutes
      </span>
      <input
        class="remember"
        name="remember"
        type="checkbox"
      />
      <span />
    </label>
    <p
      class="buttonsLine"
    >
      <a
        class="cancel"
        href="/"
      >
        Cancel
      </a>
      <button
        class="submit"
        type="submit"
      >
        Sign
      </button>
      <output
        class="errorTooltip"
        hidden=""
      >
        You need at least 0.0100 KILTs to pay the fee
      </output>
    </p>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </form>
</div>
`;

exports[`DidDowngrade promo used for web3name but not DID 1`] = `
<div>
  <form
    autocomplete="off"
    class="container"
  >
    <h1
      class="heading"
    >
      Remove DID
    </h1>
    <p
      class="subline"
    >
      Remove DID from the KILT blockchain
    </p>
    <section
      class="container"
    >
      <div
        class="avatar fullDid"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Identity with on-chain DID and lots of coins
        </span>
      </div>
    </section>
    <p
      class="costs"
    >
      Refund Amount: 
      <span
        class="amount"
      >
        0
        <span
          class="fractional"
        >
          .9900
        </span>
         
        <span
          aria-label="KILT Coin"
          class="coin"
        />
      </span>
    </p>
    <p
      class="details"
    >
       Deposit 
      1.0000
       
      <span
        aria-label="KILT Coin"
        class="coin"
      />
       - fee 
      0.0100
       
      <span
        aria-label="KILT Coin"
        class="coin"
      />
    </p>
    <p
      class="explanation"
    >
      Take care: this identity won’t be able to get a DID again.
    </p>
    <p
      class="resetLine"
    >
      <label
        class="passwordLabel"
        for="password"
      >
        Sign with password:
      </label>
      <a
        class="reset"
        href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo/reset"
      >
        Reset password
      </a>
    </p>
    <p
      class="passwordLine"
    >
      <input
        class="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        aria-label="Show"
        class="show"
        title="Show"
        type="button"
      />
      <output
        class="errorTooltip"
        hidden=""
      />
    </p>
    <label
      class="rememberLabel"
    >
      <span>
        Remember password for the next 15 minutes
      </span>
      <input
        class="remember"
        name="remember"
        type="checkbox"
      />
      <span />
    </label>
    <p
      class="buttonsLine"
    >
      <a
        class="cancel"
        href="/"
      >
        Cancel
      </a>
      <button
        class="submit"
        type="submit"
      >
        Sign
      </button>
      <output
        class="errorTooltip"
        hidden=""
      >
        You need at least 0.0100 KILTs to pay the fee
      </output>
    </p>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </form>
</div>
`;
